tst/scripts/utils/artemis.py

```diff
@@ -122,6 +146,131 @@ def run(nproc, input_filename, arguments, restart=None):
+
+        offset = f["LogicalLocations"][...][ind, :]
+        x = f["Locations/x"][...][ind, :]
+        y = f["Locations/y"][...][ind, :]
+        z = f["Locations/z"][...][ind, :]
+        db = f["gas.prim.density_0"][...][ind, :]
+        pb = f["gas.prim.pressure_0"][...][ind, :]
+        ub = f["gas.prim.velocity_0"][...][ind, 0, :]
+        vb = f["gas.prim.velocity_0"][...][ind, 1, :]
+        wb = f["gas.prim.velocity_0"][...][ind, 2, :]
+        sys = f["Params"].attrs["artemis/coord_sys"]
+        time = f["Info"].attrs["Time"]
+
+    nb, bsz, bsy, bsx = db.shape
+
+    offset[:, 0] -= offset[:, 0].min()
+    offset[:, 1] -= offset[:, 1].min()
+    offset[:, 2] -= offset[:, 2].min()
+
+    nx = bsx * (offset[:, 0].max() + 1)
+    ny = bsy * (offset[:, 1].max() + 1)
+    nz = bsz * (offset[:, 2].max() + 1)
+
+    d = np.zeros((nz, ny, nx))
+    p = np.zeros((nz, ny, nx))
+    u = np.zeros((nz, ny, nx))
+    v = np.zeros((nz, ny, nx))
+    w = np.zeros((nz, ny, nx))
+    for n in range(nb):
+        ix = offset[n, 0]
+        iy = offset[n, 1]
+        iz = offset[n, 2]
+        d[
+            iz * bsz : (iz + 1) * bsz,
+            iy * bsy : (iy + 1) * bsy,
+            ix * bsx : (ix + 1) * bsx,
+        ] = db[n, :, :, :]
+        p[
+            iz * bsz : (iz + 1) * bsz,
+            iy * bsy : (iy + 1) * bsy,
+            ix * bsx : (ix + 1) * bsx,
+        ] = pb[n, :, :, :]
+        u[
+            iz * bsz : (iz + 1) * bsz,
+            iy * bsy : (iy + 1) * bsy,
+            ix * bsx : (ix + 1) * bsx,
+        ] = ub[n, :, :, :]
+        v[
+            iz * bsz : (iz + 1) * bsz,
+            iy * bsy : (iy + 1) * bsy,
+            ix * bsx : (ix + 1) * bsx,
+        ] = vb[n, :, :, :]
+        w[
+            iz * bsz : (iz + 1) * bsz,
+            iy * bsy : (iy + 1) * bsy,
+            ix * bsx : (ix + 1) * bsx,
+        ] = wb[n, :, :, :]
+
+    x = np.linspace(x.min(), x.max(), nx + 1)
+    y = np.linspace(y.min(), y.max(), ny + 1)
+    z = np.linspace(z.min(), z.max(), nz + 1)
+    return time, x, y, z, [d, u, v, w, p / d]
+
+
+# Associated functions for plotting Artemis data
+def create_colorbar(ax, norm, where="top", cax=None, cmap="viridis", **kargs):
+    import matplotlib
+    import matplotlib.cm
+    from mpl_toolkits.axes_grid1 import make_axes_locatable
+
+    labelsize = kargs.pop("labelsize", 14)
+    labelsize = kargs.pop("fontsize", 14)
+
+    if cax is None:
+        divider = make_axes_locatable(ax)
+        cax = divider.append_axes(where, size="3%", pad=0.05)
+
+    cmap = matplotlib.cm.get_cmap(cmap)
+    cb = matplotlib.colorbar.ColorbarBase(
+        ax=cax, cmap=cmap, norm=norm, orientation="horizontal", **kargs
+    )
+    cb.ax.xaxis.set_ticks_position(where)
+    cb.ax.xaxis.set_label_position(where)
+    cb.ax.tick_params(labelsize=labelsize)
+
+    return cb
+
+
+# Shared function to compute analytic Ogilvie & Lubow (2002) spiral positions
+def spiral_pos(r, r0=1.0, p0=np.pi, h=0.05):
```

Comment:
This doesn't belong in `artemis.py`. It belongs in `binary.py`, and other files can import `binary` to use it if they want it.

Reply:
One test should not load another test (with the exception of derived tests).  

Moving this to a shared place prevents code duplication. 

I am open to moving this somewhere besides `artemis.py`, but not having e.g., `disk_nbody` load `binary`.

Reply:
I wouldn't call 
```
from binary import spiral_pos
```
loading the test, but loading an analysis function

Reply:
Rather than putting test specific things in something like `utils/analysis.py`, I would prefer to restructure the tests a bit. For example, all tests of binaries go in a `tst/scripts/binary/` directory and they all can share an `../binary/analysis.py` file with binary-specific functions. Or all tests using a specific pgen go in `tst/scripts/pgen` (I slightly prefer this).

Reply:
I think a general hierarchical model like
```
tst/scripts/
  problemA/
    problemA_var1.py
    problemB_var2.py
    problemA_analysis.py
  problemB/
    problemB.py
  utils/
    analysis.py // routines used by both problemA and problemB
```
is fine, I don't have a strong preference for names either exactly matching or not exactly matching pgen names, but I do think shared utility functions should 1) live at a directory level common to all tests that use them 2) have unique names between different analysis scripts so we dont run into issues with `import analysis` all over the place confusing python

Reply:
> tst/scripts/utils/analysis.py

Open to this.  

> Rather than putting test specific things in something like utils/analysis.py, I would prefer to restructure the tests a bit. For example, all tests of binaries go in a tst/scripts/binary/ directory and they all can share an ../binary/analysis.py file with binary-specific functions. Or all tests using a specific pgen go in tst/scripts/pgen (I slightly prefer this).

Or this.

Reply:
Especially if moving to a hierarchy of analysis routines, should we even have a `utils` dir? `utils/analysis.py` and `utils/artemis.py` could just live in `tst/scripts`?

Reply:
For this MR, can we stick with moving these shared functions to a shared `tst/scripts/utils/analysis.py`? I am not opposed to the hierarchical approach, but this (1) will change how we glob for tests (we will need to be careful here...) and (2) will change how we import.  It is probably worth thinking through where the analysis scripts live too, as I like the current design where every `.py` file in the `scripts` subdirectory is a test (well... with the exception of `utils`).  Maybe something like `pgen_name/analysis/analysis.py`? This might make the `glob` easier too.  

I already sunk more time into this PR than you probably wanted me to 😅 And even though the analysis treatment is not finalized, it did eliminate significant code duplication.

Reply:
Yeah sorry I should've added here, I don't think this is a change for this PR.